### ISSUE TITLE
FFTW3 for TI c6x DSPs

### ISSUE THREAD

**Issue by kmpape** [FFTW/fftw3#199]
Hi all,

I am working on the **TI c6678 DSP** and struggling to find a suitable FFT routine. They either offer a library for FFTs with length 2^x (zero-padding not possible in my case) or a library that includes OpenMP for FFTs with length >= 1024. I did some timings and the performance without using SIMD and proper alignment is extremely bad. I had the same experience with some matrix-vector multiplications. I am certain that there is no problem with the cache and I did these timings placing code & data in the memory closest to the core. Before diving into porting to the DSP I would like to ask some questions.

- Does somebody know of someone who ported FFTW3 to a TI c6x? I have checked google and TI's developer forum (on which several people requested this already actually) but found nothing
- I do have some experience with SIMD on the c6678 but after having a brief look at the SIMD folder here I realized that several things are called differently. Would there be any way to get some assistance with this? I can do the work but I would certainly need some assistance if I am stuck.

I believe that there would be some interest for having a proper FFT routine like FFTW3 on the TI c6x DSPs. I work on the c6678 but the SIMDs are all the same (more or less) for the c6x series.

Best wishes,
Idris

In case someone needs it, the following is to be pasted into kernel/cycle.h. Careful though as the counter might overflow.
`#if defined(SOC_C6678)`
`#include <c6x.h>`
`typedef long long int ticks;`
`static inline void init_TSC(void)`
`{`
`    TSCL = 0; // write any value to start cycle counter`
`}`
`static inline ticks getticks(void)`
`{`
`  return (ticks)_itoll(TSCH, TSCL);`
`}`
`INLINE_ELAPSED(inline)`
`#define HAVE_TICK_COUNTER`
`#endif`

**Comment by rdolbeau** [comment]
The SIMD you need to modify is in simd-support ; the simd directories in rdft and dft only need minimal changes (they're mostly auto-generated codelets).
You can look at the branches [arm-sve-clean](https://github.com/rdolbeau/fftw3/tree/arm-sve-clean) and [riscv-v-clean](https://github.com/rdolbeau/fftw3/tree/riscv-v-clean) on my fork for examples of what's currently needed to support a new SIMD ISA. Those branches are likely more complex than needed as they have multiple sets of codelets of various width for those scalable SIMD ISA. Fixed-width will just need one set of codelets, like for others in simd-support. Start by looking at one with a similar width (SSE, AltiVec and NEON are 128, AVX is 256, AVX-512 is 512).
GCC has support for 'generic simd', if it can generates the proper instructions for your DSP you might try the 'generic simd' targets as well.

**Comment by stevengj** [comment]
Note that FFTW was never really intended for small embedded systems and DSP chips, but maybe DSP chips have gotten a lot more beefy these days?

**Comment by hominhquan** [comment]
Porting CPU-legacy libraries on embedded/DSP processors is a touchy task. 

- My experience with porting FFTW on our company Kalray MPPA arch is that some default SIMD `codelets` in FFTW are very heavy in **code-size** and **stack-size**, that can be performance stopper for embedded/DSP. They are for example (just to cite some) `n1fv_128.c, n1fv_64.c,  n2fv_64.c, t1fv_64.c, q1fv_8.c etc...`.

- Luckily, I think FFTW offers possibility to disable them (e.g. in `dft/simd/<arch>/codlist.mk`) and manages to fallback to others (smaller) codelets. @stevengj Can you confirm this ?

**Comment by stevengj** [comment]
Yes, you can certainly customize the list of codelets if you want.

(In generally, FFTW was designed to tradeoff code size for speed and flexibility, so it is a pretty heavyweight dependency on embedded systems.  On the other hand, embedded systems have gotten a lot bigger since 1997.)
